Maple2.Server.Game/Manager/ShopManager.cs

```diff
@@ -0,0 +1,609 @@
+        // Follows a time schedule
+        DateTime now = DateTime.Now;
+        switch (interval) {
+            case ShopRestockInterval.Minute:
+                return new DateTime(now.Year, now.Month, now.Day, now.Hour, now.Minute, 0).AddMinutes(1).ToEpochSeconds();
+            case ShopRestockInterval.Day:
+                return new DateTime(now.Year, now.Month, now.Day, 0, 0, 0).AddDays(1).ToEpochSeconds();
+            case ShopRestockInterval.Week:
+                return DateTime.Now.NextDayOfWeek(Constant.ResetDay).Date.ToEpochSeconds();
+            case ShopRestockInterval.Month:
+                return new DateTime(now.Year, now.Month, 1, 0, 0, 0).AddMonths(1).ToEpochSeconds();
+            default:
+                logger.Error("Unknown restock interval {Interval}", interval);
+                return long.MaxValue;
+        }
+    }
+
+    private long GetInstantRestockTime(ShopRestockInterval interval) {
+        return interval switch {
+            ShopRestockInterval.Minute => DateTime.Now.AddMinutes(1).ToEpochSeconds(),
+            ShopRestockInterval.Day => DateTime.Now.AddDays(1).ToEpochSeconds(),
+            ShopRestockInterval.Week => DateTime.Now.AddDays(7).ToEpochSeconds(),
+            ShopRestockInterval.Month => DateTime.Now.AddMonths(1).ToEpochSeconds(),
+            _ => long.MaxValue,
+        };
+    }
+
+
+    public void Buy(int shopItemId, int quantity) {
+        if (activeShop == null) {
+            return;
+        }
+
+        if (!activeShop.Items.TryGetValue(shopItemId, out ShopItem? shopItem)) {
+            session.Send(ShopPacket.Error(ShopError.s_err_invalid_item));
+            return;
+        }
+
+
+        if (shopItem.RestrictedBuyData != null) {
+            RestrictedBuyData buyData = shopItem.RestrictedBuyData;
+            if (DateTime.Now.ToEpochSeconds() < buyData.StartTime || DateTime.Now.ToEpochSeconds() > buyData.EndTime) {
+                session.Send(ShopPacket.Error(ShopError.s_err_invalid_item_cannot_buy_by_period));
+                return;
+            }
+
+            if (buyData.TimeRanges.Count > 0) {
+                int totalSeconds = (int) new TimeSpan(DateTime.Now.Hour, DateTime.Now.Minute, DateTime.Now.Second).TotalSeconds;
+                if (!buyData.TimeRanges.Any(time => time.StartTimeOfDay > totalSeconds || time.EndTimeOfDay < totalSeconds)) {
+                    session.Send(ShopPacket.Error(ShopError.s_err_invalid_item_cannot_buy_by_period));
+                    return;
+                }
+            }
+
+            if (buyData.Days.Count > 0) {
+                if (!buyData.Days.Contains(ToShopBuyDay(DateTime.Now.DayOfWeek))) {
+                    session.Send(ShopPacket.Error(ShopError.s_err_invalid_item_cannot_buy_by_period));
+                    return;
+                }
+            }
+        }
+
+        if (shopItem.StockCount > 0 && shopItem.StockPurchased + quantity > shopItem.StockCount) {
```

Comment:
I'm worried adding `shopItem.StockPurchased + quantity` given `quantity` is user input, you can cause overflow. I think a better check is something like `quantity > shopItem.StockCount - shopItem.StockPurchased`